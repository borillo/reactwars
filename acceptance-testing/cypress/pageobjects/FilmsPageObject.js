class FilmsPageObject {
  constructor(wrapper) {
    this.wrapper = wrapper;
  }

  open() {
    this.wrapper.server();
    this.wrapper.route("/api/films", "fixture:films");

    this.wrapper.visit("/films");
  }

  title() {
    return this.wrapper.get(".films-panel h1").first();
  }

  films() {
    return this.wrapper.get(".film");
  }

  filmsTitles() {
    return this.wrapper.get(".film .title");
  }

  filmsEpisodes() {
    return this.wrapper.get(".film .episode");
  }

  numVotes(filmName) {
    return this.wrapper
      .contains(".film", filmName)
      .find(".votes")
      .invoke("text")
      .then(text => {
        if (!text) return 0;
        return parseInt(text, 10);
      });
  }

  voteFilm(filmName) {
    this.wrapper
      .contains(filmName)
      .parentsUntil(".film")
      .parent()
      .find(".vote")
      .click();
  }

  registerFilm(film) {
    return this.wrapper
      .get(".new-film-title")
      .type(film.title)
      .get(".new-film-director")
      .type(film.director)
      .get(".new-film-year")
      .type(film.year)
      .get(".new-film-poster")
      .type(film.poster)
      .get(".save-new-film")
      .click();
  }
}

export default FilmsPageObject;
