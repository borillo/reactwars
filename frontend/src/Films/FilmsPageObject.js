export default class FilmsPageObject {
  constructor({ getByTestId, container, debug }) {
    this.getByTestId = getByTestId;
    this.container = container;
    this.debug = debug;
  }

  retrieveDom() {
    return this.container;
  }

  obtainFilms() {
    return this.getByTestId("films").querySelectorAll(".film");
  }

  obtainFilmsTitles() {
    const nodeTitles = this.container.querySelectorAll(".title");

    return [...nodeTitles].map(node => node.textContent);
  }

  obtainFilmsEpisodes() {
    const nodeEpisodes = this.container.querySelectorAll(".episode");

    return [...nodeEpisodes].map(node => node.textContent);
  }
}
