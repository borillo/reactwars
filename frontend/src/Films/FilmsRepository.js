export default {
  retrieveFilms() {
    return fetch("/api/films").then(response => response.json());
  }
};
