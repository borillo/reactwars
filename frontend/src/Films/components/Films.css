.films {
  display: flex;
}

.films .films-list {
  min-width: 27rem;
}

.films-panel input {
  width: 25rem;
}
