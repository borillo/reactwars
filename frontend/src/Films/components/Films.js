import React from "react";
import PropTypes from "prop-types";

import Film from "./Film";

const Films = ({ data: films }) => {
  return (
    <div className="films" data-testid="films">
      {films.map(film => (
        <Film key={film.episode} title={film.title} episode={film.episode} />
      ))}
    </div>
  );
};

Films.propTypes = {
  data: PropTypes.array.isRequired
};

export default Films;
