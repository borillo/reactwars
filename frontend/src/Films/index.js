import { connect } from "react-redux";

import { loadFilmsAction } from "./actions/creators";

import FilmsPanel from "./components/FilmsPanel";

const mapStateToProps = ({ films: data }) => ({ data });

const mapDispatchToProps = dispatch => {
  return {
    loadFilms: () => {
      dispatch(loadFilmsAction());
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(FilmsPanel);
