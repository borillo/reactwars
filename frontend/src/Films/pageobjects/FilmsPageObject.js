export default class FilmsPageObject {
  constructor({ getByTestId, container }) {
    this.getByTestId = getByTestId;
    this.container = container;
  }

  retrieveDom() {
    return this.container;
  }
  
  obtainFilms() {
    return this.getByTestId("films").children;
  }

  obtainFilmsTitles() {
    return []; /*this.wrapper
      .update()
      .find(".film .title")
      .map(film => film.text());*/
  }

  obtainFilmsEpisodes() {
    return []; /*this.wrapper
      .update()
      .find(".film .episode")
      .map(film => film.text());*/
  }
}
