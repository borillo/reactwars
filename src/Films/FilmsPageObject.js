export default class FilmsPageObject {
  constructor(wrapper) {
    this.wrapper = wrapper;
  }
 
  obtainFilms() {
    return this.wrapper.update().find('.film');
  }
  
  obtainFilmsTitles() {
    return this.wrapper.update().find('.film .title').map((film) => film.text());
  }

  obtainFilmsEpisodes() {
    return this.wrapper.update().find('.film .episode').map((film) => film.text());
  }
}
