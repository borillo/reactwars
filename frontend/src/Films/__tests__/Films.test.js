jest.mock("../repositories/FilmsRepository", () => {
  return {
    retrieveFilms() {
      return Promise.resolve([{ episode: 1, title: "A" }]);
    }
  };
});

import React from "react";
import { Provider } from "react-redux";

import { createStore, combineReducers, applyMiddleware } from "redux";
import thunkMiddleware from "redux-thunk";

import { render } from "react-testing-library";

import Films from "../index";
import FilmsPageObject from "../pageobjects/FilmsPageObject";

import filmsReducers from "../actions/reducers";

const configureStore = () => {
  return createStore(
    combineReducers({
      films: filmsReducers
    }),
    applyMiddleware(thunkMiddleware)
  );
};

describe("Films", () => {
  let pageObject;

  beforeEach(() => {
    const store = configureStore();

    const componentDom = render(
      <Provider store={store}>
        <Films />
      </Provider>
    );

    pageObject = new FilmsPageObject(componentDom);
  });

  test("should be listed", async () => {
    let filmList = pageObject.obtainFilms();

    expect(filmList).toHaveLength(1);
  });

  /*it("should show name and episode number", async () => {
    const filmsTitles = films.obtainFilmsTitles();
    expect(filmsTitles).toEqual(FILM_TITLES);

    const filmsEpisodes = films.obtainFilmsEpisodes();
    expect(filmsEpisodes).toEqual(FILM_EPISODES);
  });

  it("should be ordered by episode number", async () => {
    let episodes = films.obtainFilmsEpisodes();
    episodes.forEach((episode, index) => {
      expect(episode).toEqual(`Episode ${index + 1}`);
    });
  });*/

  it("should match snapshot", async () => {
    expect(pageObject.retrieveDom()).toMatchSnapshot();
  });
});
