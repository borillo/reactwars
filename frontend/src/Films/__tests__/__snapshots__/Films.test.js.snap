// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Films should match snapshot 1`] = `
<div>
  <div
    class="films-panel"
  >
    <h1>
      Film list
    </h1>
    <div
      class="films"
      data-testid="films"
    >
      <div
        class="film"
      >
        <div
          class="title"
        >
          <strong>
            A
          </strong>
        </div>
        <div
          class="episode"
        >
          Episode 
          1
        </div>
      </div>
    </div>
  </div>
</div>
`;
